File: BorderlessWindow/MainArea.py
from PySide6.QtCore import Qt
from PySide6.QtGui import QPaintEvent, QPainter, QColor
from PySide6.QtWidgets import QVBoxLayout, QWidget

from TitleBar import TitleBar


class CustomerAreaWidget(QWidget):
    """客户区"""

    def __init__(self, parent):
        super().__init__(parent=parent)

    def paintEvent(self, event: QPaintEvent):
        painter = QPainter(self)
        painter.setPen(Qt.PenStyle.NoPen)
        painter.setBrush(QColor(240, 240, 240))
        painter.drawRect(self.rect())


class MainArea(QWidget):
    """主区域，包含客户区和标题栏"""

    def __init__(self, parent: QWidget):
        super().__init__(parent)
        self.verticalLayout = QVBoxLayout(self)
        self.verticalLayout.setContentsMargins(0, 0, 0, 0)
        self.verticalLayout.setSpacing(0)

        self.title_bar = TitleBar(self)
        self.customer_area = CustomerAreaWidget(self)

        self.verticalLayout.addWidget(self.title_bar)
        self.verticalLayout.addWidget(self.customer_area)

    def TitleBar(self):
        return self.title_bar

File: BorderlessWindow/mainarea.cpp
#include "mainarea.h"

MainArea::MainArea(QWidget *parent)
    : QWidget{parent},
      vertical_layout{new QVBoxLayout(this)},
      title_bar{new StandardTitleBar(this)},
      customer_area_widget{new QWidget(this)}
{
    this->vertical_layout->setContentsMargins(0, 0, 0, 0);
    this->vertical_layout->setSpacing(0);
    this->vertical_layout->addWidget(this->title_bar);
    this->vertical_layout->addWidget(this->customer_area_widget);
}

QWidget *MainArea::customerAreaWidget()
{
    return this->customer_area_widget;
}

StandardTitleBar *MainArea::titleBar()
{
    return this->title_bar;
}

void MainArea::paintEvent(QPaintEvent *event)
{
    QWidget::paintEvent(event);
    QPainter painter = QPainter(this);
    painter.setRenderHint(QPainter::RenderHint::Antialiasing);

    painter.setBrush(QColor(240, 240, 240));

    if (!this->window()->isMaximized())
    {
        const_cast<QPen&>((painter.pen())).setWidth(0);
        painter.drawRoundedRect(1,1,width()-2,height()-2,10,10);
    }
    else if (this->window()->isMaximized())
    {
        painter.setPen(Qt::PenStyle::NoPen);
        painter.drawRect(this->rect());
    }
}
